src/main/scala/com/fulcrumgenomics/util/DemuxFastqs.scala

```diff
@@ -0,0 +1,538 @@
+    }
+  }
+
+  /** Reads multiple FASTQs at the same time, ensuring their read names match along the way.  This is
+    * useful for when we don't have a single interleaved FASTQ. */
+  private[util] class ParallelFastqReader(fastq: List[Path], fastqs: List[Path]*) extends Iterator[List[FastqRecord]] {
+
+    val readers: List[SequentialFastqReader] = (fastq :: fastqs.toList).map { otherFastqs =>
+      if (otherFastqs.size != fastq.size) throw new IllegalArgumentException("List of list fastqs had differing lengths.")
+      new SequentialFastqReader(otherFastqs)
+    }
+
+    def next: List[FastqRecord] = {
+      val records = readers.map(_.next)
+      verifySameReadNames(records)
+      records
+    }
+
+    def hasNext: Boolean = {
+      if (readersHaveNext(true)) true
+      else if (readersHaveNext(false)) false
+      else throw new IllegalArgumentException("Fastqs have differing lengths")
+    }
+
+    private def readersHaveNext(desiredValue: Boolean): Boolean = !readers.exists(_.hasNext != desiredValue)
+
+    private def verifySameReadNames(records: List[FastqRecord]): Unit = {
+      val readName: String = extractName(records.head.getReadHeader)
+      records.drop(0).foreach { record =>
+        val otherReadName: String = extractName(record.getReadHeader)
+        if (readName.compareTo(otherReadName) != 0) throw new IllegalArgumentException(String.format("Mismatching read names in FASTQS:\n%s\n%s\n", readName, otherReadName))
+      }
+    }
+
+    def close() = readers.foreach(_.close())
+  }
+
+  private class ReadStructureInfo(readStructureString: String) extends ReadStructure(readStructureString) {
+    val templateCycles: Array[Int]         = this.templates.getCycles
+    val sampleBarcodeCycles: Array[Int]    = this.sampleBarcodes.getCycles
+    val molecularBarcodeCycles: Array[Int] = this.molecularBarcode.getCycles
+  }
+
+  /** Helper class to convert fastq records into SAM records and write them out.
+    *
+    * @param sampleSheet the sample sheet describing all the samples
+    * @param readStructureInfos the read structures info, one for each expected fastq record that will be passed
+    *                           into `convertFastqsAndWriteRecords`.
+    */
+  private class FastqConverter(val sampleSheet: SampleSheet,
+                               val umiTag: String,
+                               val qualityFormat: FastqQualityFormat,
+                               val minQ: Int,
+                               val maxQ: Int,
+                               val maxMismatches: Int,
+                               val minMismatchDelta: Int,
+                               val maxNoCalls: Int,
+                               val readStructureInfos: ReadStructureInfo*) {
+    /** Sample barcode to its associate metric */
+    private val barcodeToMetrics: mutable.Map[String, ExtractIlluminaBarcodes.BarcodeMetric] = new mutable.LinkedHashMap[String, ExtractIlluminaBarcodes.BarcodeMetric]()
```

Comment:
it would be nice to get rid of our reliance on Picard by having our own metric class.